sklearn/decomposition/_sparse_pca.py

```diff
@@ -29,6 +29,66 @@ def _check_normalize_components(normalize_components, estimator_name):
             )
 
 
+def _get_explained_variance(X, components):
+    '''Get the explained variance.
+
+    Get the explained variance from the principal components of the
+    data. This follows the method outlined in [1] section 3.4 (Adjusted Total
+    Variance). For an alternate approach (not implemented here), see [2].
+
+    Parameters
+    ----------
+    X : ndarray, shape (n_samples, n_features)
+        The feature vector. n_samples and n_features are the number of
+        samples and features, respectively.
+
+    components : array, shape (n_components, n_features)
+        The (un-normalized) principle components. [1]
+
+    Notes
+    -----
+    The variance ratio may not be computed. The main reason is that we
+    do not know what the total variance is since we did not compute all
+    the components.
+    Orthogonality is enforced in this case. Other variants exist that don't
+    enforce this [2].
+
+    References
+    ----------
+    .. [1] Journal of Computational and Graphical Statistics, Volume 15, Number
+        2, Pages 265–286. DOI: 10.1198/106186006X113430
+    .. [2] Rodolphe Jenatton, Guillaume Obozinski, Francis Bach ; Proceedings
+        of the Thirteenth International Conference on Artificial Intelligence
+        and Statistics, PMLR 9:366-373, 2010.
+    '''
+    # the number of samples
+    n_samples = X.shape[0]
+    n_components = components.shape[0]
+    unit_vecs = components.copy()
+    components_norm = np.linalg.norm(components, axis=1)[:, np.newaxis]
+    components_norm[components_norm == 0] = 1
+    unit_vecs /= components_norm
+
+    # Algorithm, as we compute the adjustd variance for each component, we
+    # subtract the variance from components in the direction of previous axes
+    proj_corrected_vecs = np.zeros_like(components)
+    for i in range(n_components):
+        vec = components[i].copy()
+        # subtract the previous projections
+        for j in range(i):
+            vec -= np.dot(unit_vecs[j], vec)*unit_vecs[j]
```

Comment:
@Batalex I have doubts that what you do here is what is described in "section 3.4 ADJUSTED TOTAL VARIANCE" of the referenced paper. You're basically regressing out one component at a time but since the component are not orthogonal you can put back in vec some variance from previous components at each stage. You need to do an actually projection. It's a bit like the difference between matching pursuit and orthogonal matching pursuit.

can you confirm?

Reply:
To be honest this PR is a continuation of  #11527 and I am struggling with the content of the paper so the reason for the original implementation is eluding me.

Nonetheless I found this Cross Validated post which propose an implementation for the explained variance: https://stats.stackexchange.com/questions/438508/what-does-sparse-pca-implementation-in-python-do

Is this what you are looking for?